Nativize summarize

patrick-schultz opened the pull request.

patrick-schultz commented on src/main/scala/is/hail/variant/VariantDataset.scala:
```diff
@@ -163,8 +163,10 @@ g = let
+    val rowType = vds.rowType
+    vds.rdd2.mapPartitions { it =>
+      Iterator.single(it.foldLeft(new SummaryCombiner(rowType))(_.merge(_)))
+    }.reduce(_.merge(_))
```
Is there a better pattern for doing an aggregation with per-partition allocations?

cseed commented on src/main/scala/is/hail/utils/SummaryCombiner.scala:
```diff
@@ -39,28 +41,49 @@ class SummaryCombiner[T](f: Iterable[T] => Int) extends Serializable {
 
-  def merge(data: (Variant, (Annotation, Iterable[T]))): SummaryCombiner[T] = {
+  def merge(rv: RegionValue): SummaryCombiner = {
+    require(rowType.fieldType(1).isInstanceOf[TVariant])
```
This is unnecessary since the next line will immediately fail with a class cast exception if it isn't a `TVariant`.

cseed replied:
I think the readability of this code would be greatly improved by a `VariantView`.  Luckily @danking already has a start on that: https://github.com/hail-is/hail/pull/2428.  Can you coordinate with him to extend it to handle this use case?

cseed commented on src/main/scala/is/hail/variant/HTSGenotypeView.scala:
```diff
@@ -290,11 +290,13 @@ object HardCallView {
 }
 
-abstract class HardCallView {
+abstract class HardCallView extends Serializable {
```
I think I'm against making this serializable.  See my comment above.

cseed commented on src/main/scala/is/hail/utils/SummaryCombiner.scala:
```diff
@@ -39,28 +41,49 @@ class SummaryCombiner[T](f: Iterable[T] => Int) extends Serializable {
     }
 
-    nCalled += f(data._2._2)
+    view.setRegion(rv)
```
You should create the view in a mapPartitions call and pass it in to this function.  Same for the variant view.

But you're using an aggregate.  How to turn that into a mapPartitions?  Something like:

```
  rdd2.mapPartitions { it =>
    val comb = new SummaryCombiner
    val view = ...
    it.map { rv => comb.merge(rv, view) }
    Iterator(comb)
  }.aggregate(_.merge(_), _.merge(_))
```

danking replied:
Might be nice to have:

```scala
def rdd2Aggregate[T](setup: () => T)(seqOp: (T, RV) => (), combOp (T, T) => (T)): T =
  rdd2.mapPartitions { it =>
    val t = setup()
    it.map { rv => seqOp(t, rv) }
    Iterator(t)
  }.aggregate(combOp, combOp)
```

so we can do

```scala
rdd2.rdd2Aggregate(() => (new SummaryCombiner(), ...))(
  { case (c, v) => c.merge(v, view) }, _.merge(_))
```

patrick-schultz commented:
I believe I've addressed all the comments, other than waiting on `VariantView`

cseed commented:
When you fix the comments you should dismiss the review it shows up for the reviewers on the scorecard again so they know to take another look.

patrick-schultz commented:
@cseed I didn't dismiss your review because I haven't rewritten using `VariantView`. Or should that be a separate PR?

cseed commented:
Oops, you're right, I stand corrected!

patrick-schultz commented:
I used `RegionValueVariant` to clean up some of the code, and fixed a couple of bugs from #2451 in the process. I also replaced the `aggregatePartitions` method I wrote in e5f87c3 following @danking's comment, which was defined on `OrderedRDD2`, with `aggregateWithContext`, which is defined on a rich wrapper around `RDD`. I put it in the spark package to get access to private methods, making the implementation cleaner. It is now a simple modification of the implementation of `RDD.aggregate`.
